src/coprocessor/dag/expr/mod.rs

```diff
@@ -210,7 +210,7 @@ impl Expression {
     ) -> Result<Option<Cow<'a, str>>> {
         let bytes = try_opt!(self.eval_string(ctx, row));
         let chrst = self.get_tp().get_charset();
-        if charset::UTF8_CHARSETS.contains(&chrst) {
+        if charset::compatible_with_utf8(chrst) {
```

Comment:
I think we can remove this check.
`binary` is not compatible with `utf8`.

Reply:
Then why remove the check? If it's not compatible, an error should be reported for that TiKV can only handle UTF-8 string.

Reply:
I think TiKV should not return errors if TiDB doesn't.

Reply:
After we add this check, TiKV won't return errors.

Reply:
If it's a String, then TiDB should ensure that it's encoded in UTF-8. Otherwise we should consider the protocol is broken, there should be some bugs.

Reply:
There definitely some bugs since we are not same as MySQL, but It's not protocol broken. `latin1` is alllowed in TiDB right now, so when `latin1` compares an `utf8`, TiDB doesn't return error.

Reply:
The comparison you mention is done by bytes and has nothing to do with this method.

Reply:
Fixed. Now `Bin` isn't compatible with `UTF-8`. PTAL again, thanks.

Reply:
But it doesn't handle binary.

Reply:
So can we remove the `compatible_with_utf8` method now?